frontend/src/lib/v2/StaticFlow.ts

```diff
@@ -0,0 +1,362 @@
+/**
+ * Build single layer graph of a pipeline definition in Reactflow.
+ * @param pipelineSpec Full pipeline definition
+ * @param componentSpec Designated layer of a DAG/sub-DAG as part of pipelineSpec
+ * @returns Graph visualization as Reactflow elements (nodes and edges)
+ */
+function buildDag(pipelineSpec: PipelineSpec, componentSpec: ComponentSpec): Elements {
+  const dag = componentSpec.getDag();
+  if (!dag) {
+    throw new Error('dag not found in component spec.');
+  }
+
+  const componentsMap = pipelineSpec.getComponentsMap();
+  let flowGraph: FlowElement[] = [];
+
+  const tasksMap = dag.getTasksMap();
+  console.log('tasksMap count: ' + tasksMap.getLength());
+
+  addTaskNodes(tasksMap, componentsMap, flowGraph);
+  addArtifactNodes(tasksMap, componentsMap, flowGraph);
+
+  addTaskToArtifactEdges(tasksMap, componentsMap, flowGraph);
+  addArtifactToTaskEdges(tasksMap, componentsMap, flowGraph);
+  addTaskToTaskEdges(tasksMap, flowGraph);
+
+  return buildGraphLayout(flowGraph);
+}
+
+function addTaskNodes(
+  tasksMap: jspb.Map<string, PipelineTaskSpec>,
+  componentsMap: jspb.Map<string, ComponentSpec>,
+  flowGraph: FlowElement<any>[],
+) {
+  // Add tasks as nodes to the Reactflow graph.
+  tasksMap.forEach((taskSpec, taskKey) => {
+    const componentPair = getComponent(taskKey, taskSpec, componentsMap);
+    if (componentPair === undefined) {
+      return;
+    }
+    const { componentRefName, componentSpec } = componentPair;
+
+    // Component can be either an executor or subDAG,
+    // If this is executor, add the node directly.
+    // If subDAG, add a node which can represent expandable graph.
+    const name = taskSpec?.getTaskInfo()?.getName();
+    if (componentSpec.getExecutorLabel().length > 0) {
+      // executor label exists means this is a single execution node.
+      const node: Node = {
+        id: getTaskNodeKey(taskKey), // Assume that key of `tasks` in `dag` is unique.
+        data: { label: name, taskType: TaskType.EXECUTOR },
+        position: { x: 100, y: 200 },
+        // TODO(zijianjoy): This node styling is temporarily.
+        style: {
+          color: 'white',
+          backgroundColor: '#00BBF9',
+          borderColor: 'transparent',
+          borderRadius: '30px',
+        },
+      };
+      flowGraph.push(node);
+    } else if (componentSpec.hasDag()) {
+      // dag exists means this is a sub-DAG instance.
+      const node: Node = {
+        id: getTaskNodeKey(taskKey),
+        data: { label: 'DAG: ' + name, taskType: TaskType.DAG },
+        position: { x: 100, y: 200 },
+        // TODO(zijianjoy): This node styling is temporarily.
+        style: {},
+      };
+      flowGraph.push(node);
+    } else {
+      console.warn('Component ' + componentRefName + ' has neither `executorLabel` nor `dag`');
+    }
+  });
+}
+
+function addArtifactNodes(
+  tasksMap: jspb.Map<string, PipelineTaskSpec>,
+  componentsMap: jspb.Map<string, ComponentSpec>,
+  flowGraph: FlowElement<any>[],
```

Comment:
for my curiosity, what's blocking us from typing the FlowElement statically?

Reply:
I haven't implemented this part yet. `T` in FlowElement<T> is for the type of `data` field in FlowElement. I haven't completely figure out how the data structure should look like for T. But let me just type it statically here, and change `any` to actual type later.